pennylane_lightning/src/util/LinearAlgebra.hpp

```diff
@@ -0,0 +1,676 @@
+    firstprivate(m, n, k)
+#endif
+        for (size_t row = 0; row < m; row++) {
+            for (size_t col = 0; col < n; col++) {
+                for (size_t blk = 0; blk < k; blk++) {
+                    m_out[row * n + col] +=
+                        m_left[row * k + blk] * m_right[col * n + blk];
+                }
+            }
+        }
+    } else {
+#if defined(_OPENMP)
+#pragma omp parallel for default(none) shared(m_left, m_right, m_out)          \
+    firstprivate(m, n, k)
+#endif
+        for (size_t row = 0; row < m; row += STRIDE) {
+            size_t i;
+            size_t j;
+            size_t l;
+            std::complex<T> t;
+            for (size_t col = 0; col < n; col += STRIDE) {
+                for (size_t blk = 0; blk < k; blk += STRIDE) {
+                    // cache-blocking:
+                    for (i = row; i < std::min(row + STRIDE, m); i++) {
+                        for (j = col; j < std::min(col + STRIDE, n); j++) {
+                            t = 0;
+                            for (l = blk; l < std::min(blk + STRIDE, k); l++) {
+                                t += m_left[i * k + l] * m_right[l * n + j];
+                            }
+                            m_out[i * n + j] += t;
+                        }
+                    }
+                }
+            }
+        }
+    }
+}
+
+/**
+ * @brief Calculates matrix-matrix product using the best avaiable method.
+ *
+ * @tparam T Floating point precision type.
+ * @param m_left Row-wise flatten matrix of shape m * k.
+ * @param m_right Row-wise flatten matrix of shape k * n.
+ * @param m_out Pre-allocated row-wise flatten matrix of shape m * n.
+ * @param m Number of rows of `m_left`.
+ * @param n Number of columns of `m_right`.
+ * @param k Number of rows of `m_right`.
+ * @param transpose If `true`, requires transposed version of `m_right`.
+ *
+ * @note Consider transpose=true, to get a better performance.
+ *  To transpose a matrix efficiently, check Util::Transpose
+ */
+template <class T>
+inline void matrixMatProd(const std::complex<T> *m_left,
+                          const std::complex<T> *m_right,
+                          std::complex<T> *m_out, size_t m, size_t n, size_t k,
+                          bool transpose = false) {
+    if (!m_out) {
+        return;
+    }
+    if constexpr (USE_CBLAS) {
+        constexpr std::complex<T> co{1, 0};
+        constexpr std::complex<T> cz{0, 0};
+        const auto tr = (transpose) ? CblasTrans : CblasNoTrans;
+        if constexpr (std::is_same_v<T, float>) {
+            cblas_cgemm(CblasRowMajor, CblasNoTrans, tr, m, n, k, &co, m_left,
+                        k, m_right, transpose ? n : k, &cz, m_out, n);
+        } else if constexpr (std::is_same_v<T, double>) {
+            cblas_zgemm(CblasRowMajor, CblasNoTrans, tr, m, n, k, &co, m_left,
+                        k, m_right, transpose ? n : k, &cz, m_out, n);
+        }
```

Comment:
FYI, the previous code was 
https://github.com/PennyLaneAI/pennylane-lightning/blob/8b9d4127831a1bbe18b19fe5029c3bd2707b3a0b/pennylane_lightning/src/util/Util.hpp#L973-L979